### PR TITLE
feat(gossipsub)!: return typed error from config builder

### LINKED ISSUE #4412: gossipsub: make `Err` of config builder a proper error
Currently, this is returning `&'static str`, it should return a proper `std::error::Error` instead.

### PR DESCRIPTION
**Pull request by chirag-bgh**
## Description
Closes #4412.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/gossipsub/src/error.rs [hunk 1]
```diff
@@ -120,3 +120,37 @@ impl From<std::io::Error> for PublishError {
         PublishError::TransformFailed(error)
     }
 }
+
+/// Error associated with Config building.
+#[derive(Debug)]
+pub enum ConfigBuilderError {
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/error.rs:126]: Is this re-exported in `lib.rs`?

**Review by thomaseizinger — COMMENTED** [review]
Thank you!

This is a breaking change so we'll queue it for the next release. I'll thus mark it is draft so that it doesn't get merged accidentally.

Can you please bump the version of `libp2p-gossipsub` accordingly and also write a changelog entry? See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#write-changelog-entries-for-user-facing-changes for details.

File: protocols/gossipsub/src/error.rs (hunk 1 above)
Comment (chirag-bgh) [reply protocols/gossipsub/src/error.rs:126]: No its not.

File: protocols/gossipsub/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,10 @@
+## 0.45.2 - unreleased
+
+- return typed error from config builder.
```
Comment (thomaseizinger) [inline protocols/gossipsub/CHANGELOG.md]:
```suggestion
- Return typed error from config builder.
```

File: protocols/gossipsub/Cargo.toml [hunk 3]
```diff
@@ -3,7 +3,7 @@ name = "libp2p-gossipsub"
 edition = "2021"
 rust-version = { workspace = true }
 description = "Gossipsub protocol for libp2p"
-version = "0.45.1"
+version = "0.45.2"
```
Comment (thomaseizinger) [inline protocols/gossipsub/Cargo.toml]: As mentioned, this is a breaking change so we need to bump to `0.46.0` :)

File: protocols/gossipsub/src/error.rs (hunk 1 above)
Comment (thomaseizinger) [reply protocols/gossipsub/src/error.rs:126]: It should be because it is part of the public API!

**Review by thomaseizinger — COMMENTED** [review]
Thanks!

Some more comments.

File: protocols/gossipsub/src/error.rs (hunk 1 above)
Comment (chirag-bgh) [reply protocols/gossipsub/src/error.rs:126]: Okay

**Review by thomaseizinger — APPROVED** [review]

**Comment by thomaseizinger** [comment]
Thank you! I queued this into the next milestone. It will likely still be a while until we make another release with breaking changes :)

**Comment by mxinden** [comment]
Thanks @chirag-bgh for your work!

**Pull request merged**